File: ios/OakApp/Services/ChatService.swift
import Foundation
import UIKit

/// One chat turn → a live `SSEEvent` stream (component-design.md "Services layer";
/// chat-experience.md M-CHAT-US-1/4). View models depend on this **protocol**
/// (never `LiveChatService`) so they unit-test against `FakeChatService`.
///
/// The method is **synchronous** and returns an `AsyncThrowingStream<SSEEvent, Error>`
/// immediately: it runs in the caller's context (so passing the main-actor-held
/// `[UIImage]` is safe — there is no actor hop), then hands back the stream the
/// view-model's reducer consumes. Pre-stream HTTP failures (rate limit, 413, 503…)
/// surface as a thrown `OakError` from the stream before any event is yielded; an
/// in-band SSE `error` event (transport fault) arrives as `.error(...)`. Every
/// in-domain failure rides a normal `.answer` event whose `OakAnswer.status` carries
/// the failure — never the `error` channel (sse-types.ts).
protocol ChatService: Sendable {
  /// Opens the chat stream for one turn.
  ///
  /// - `sessionId`: the client thread UUID (equals the conversation id on resume).
  /// - `message`: 0–2000 chars; MAY be empty when `images` are attached.
  /// - `images`: attached photos (≤4), encoded to raw base64 with the client-side
  ///   caps enforced by `ImageEncoder`. A cap/type violation finishes the returned
  ///   stream by throwing `OakError.imageRejected(...)` before any event is yielded.
  /// - `scopeSeed`: an explicit scope pick from the header chip, sent as
  ///   `scope_seed`; `nil` ⇒ no pick (server precedence resolves the scope). Scope
  ///   is otherwise server-controlled — the model never sees it as a tool input.
  ///
  /// Saved teams are referenced **by name in chat** (resolved server-side via
  /// `list_teams` / `get_team`), so the body carries no team id.
  func send(
    sessionId: String,
    message: String,
    images: [UIImage],
    scopeSeed: Format?
  ) -> AsyncThrowingStream<SSEEvent, Error>

  /// Reattaches to a durable turn's live stream (`GET /api/chat/turns/:id/stream`,
  /// background-turns/design.md §6.2). Yields the SAME `SSEEvent` sequence a fresh
  /// `send` would — opening with the `turn` frame, replaying the buffered events,
  /// then tailing live to the terminal `answer`/`error`/`stopped`. Used when a
  /// thread is reopened (or the app foregrounded) with a turn still generating, and
  /// to heal a mid-stream connection drop for a known `turnId`.
  ///
  /// - `turnId`: the server-minted id captured from the `turn` frame (or the
  ///   conversation's `active_turn`).
  /// - `sessionId`: the client thread UUID, sent as `?session_id=` for guest
  ///   ownership (ignored for a signed-in caller, whose Bearer identifies them).
  ///
  /// A 404 (unknown/expired turn) is thrown as `OakError.http(status: 404, …)`
  /// before any event is yielded — the caller clears the pending turn and offers a
  /// manual retry.
  func resumeStream(turnId: String, sessionId: String) -> AsyncThrowingStream<SSEEvent, Error>

  /// Explicitly stops a running turn (`POST /api/chat/turns/:id/stop`,
  /// background-turns/design.md §6.2 / BT-4). The turn is discarded server-side
  /// (nothing persisted) and subscribers receive a terminal `stopped` event. Used
  /// by the composer's Stop affordance. Stopping an already-terminal turn is a
  /// no-op; the caller tears down its local stream regardless of the result.
  func stop(turnId: String, sessionId: String) async throws
}

/// Production ``ChatService`` over ``SSEClient`` (which borrows ``OakAPIClient`` for
/// the Bearer header + base URL). A value type holding one immutable struct, so it
/// is `Sendable` without ceremony.
struct LiveChatService: ChatService {
  private let sseClient: SSEClient

  init(sseClient: SSEClient) {
    self.sseClient = sseClient
  }

  func send(
    sessionId: String,
    message: String,
    images: [UIImage],
    scopeSeed: Format?
  ) -> AsyncThrowingStream<SSEEvent, Error> {
    // Encode + validate the attached images BEFORE opening the stream (M-AC-5.5).
    // `encode` is synchronous and runs in the caller's context (the main actor),
    // so passing the main-actor-held `[UIImage]` involves no actor hop. A client
    // cap/type violation surfaces as a thrown `OakError.imageRejected(...)` from
    // the returned stream — never as a partially-attached turn.
    let encodedImages: [ChatImage]
    do {
      encodedImages = try ImageEncoder().encode(images)
    } catch let error as OakError {
      return AsyncThrowingStream { $0.finish(throwing: error) }
    } catch {
      // `encode` only throws `OakError`; map any unexpected error to a Sendable
      // `OakError` BEFORE the `@Sendable` stream closure captures it.
      let mapped = OakError.transportFailure(error)
      return AsyncThrowingStream { $0.finish(throwing: mapped) }
    }

    let request = ChatRequest(
      sessionId: sessionId,
      message: message,
      images: encodedImages.isEmpty ? nil : encodedImages,
      scopeSeed: scopeSeed
    )
    return sseClient.stream(request)
  }

  func resumeStream(turnId: String, sessionId: String) -> AsyncThrowingStream<SSEEvent, Error> {
    sseClient.resume(turnId: turnId, sessionId: sessionId)
  }

  func stop(turnId: String, sessionId: String) async throws {
    try await sseClient.stop(turnId: turnId, sessionId: sessionId)
  }
}

File: android/app/src/main/kotlin/ai/gowtam/oak/services/ChatService.kt
package ai.gowtam.oak.services

import ai.gowtam.oak.networking.OakError
import ai.gowtam.oak.networking.SseClient
import ai.gowtam.oak.wire.ChatImage
import ai.gowtam.oak.wire.ChatRecovery
import ai.gowtam.oak.wire.ChatRequest
import ai.gowtam.oak.wire.Format
import ai.gowtam.oak.wire.SseEvent
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow

/**
 * One chat turn → a live [SseEvent] stream (component-design.md "Services layer";
 * mirrors iOS `ChatService`). View models depend on this **interface** (never
 * `LiveChatService`) so they unit-test against a fake.
 *
 * [send] is synchronous and returns the [Flow] immediately: image encoding runs
 * eagerly, in the caller's context, BEFORE the stream opens. A cap/type violation
 * surfaces as a thrown [OakError.ImageRejected] the moment the returned flow is
 * collected — before any event is emitted. Pre-stream HTTP failures (rate limit,
 * 413, 503, …) surface the same way; an in-band SSE `error` event arrives as
 * [SseEvent.Error]. Every in-domain failure rides a normal [SseEvent.Answer] whose
 * `OakAnswer.status` carries the failure — never the `error` channel.
 */
interface ChatService {
    /**
     * Opens the chat stream for one turn (`POST /api/chat`), encoding [images] via
     * [ImageEncoder] before the request is sent.
     *
     * - [sessionId]: the client thread UUID (equals the conversation id on resume).
     * - [message]: 0–2000 chars; MAY be empty when [images] are attached.
     * - [images]: attached photos (≤4), encoded to raw base64 with the client-side
     *   caps enforced by [ImageEncoder]. A cap/type violation finishes the returned
     *   flow by throwing [OakError.ImageRejected].
     * - [scopeSeed]: an explicit scope pick from the header chip, sent as
     *   `scope_seed`; `null` ⇒ no pick (server precedence resolves the scope). Scope
     *   is otherwise server-controlled — the model never sees it as a tool input.
     *
     * Saved teams are referenced **by name in chat** (resolved server-side), so the
     * body carries no team id.
     */
    fun send(
        sessionId: String,
        message: String,
        images: List<SourceImage>,
        scopeSeed: Format?,
    ): Flow<SseEvent>

    /**
     * Same as [send] plus chat-qol recovery / mention fields. Default
     * implementation drops the extras so existing test doubles keep compiling.
     */
    fun send(
        sessionId: String,
        message: String,
        images: List<SourceImage>,
        scopeSeed: Format?,
        recovery: ChatRecovery?,
        mentionedTeamIds: List<String>?,
    ): Flow<SseEvent> = send(sessionId, message, images, scopeSeed)

    /**
     * Text-only convenience: opens the stream for an already-built [request] with no
     * image encode step (e.g. a resend/retry that carries no new attachments).
     */
    fun send(request: ChatRequest): Flow<SseEvent>

    /**
     * Reattaches to a durable turn's live stream
     * (`GET /api/chat/turns/:id/stream`; background-turns/design.md §4). The stream
     * replays the turn's buffered events (the client rebuilds its in-flight UI on the
     * `turn` frame) then tails live until a terminal [SseEvent]; a turn that is
     * already terminal replays through its terminal event and closes. An unknown or
     * expired turn is a pre-stream `404` → [OakError.Http] thrown before any event
     * (the caller treats it as a dead turn).
     *
     * - [turnId]: the server-minted turn id captured from the `turn` event / the
     *   conversation's `active_turn`.
     * - [sessionId]: the conversation id, the guest ownership proof (ignored for a
     *   signed-in caller, identified by the Bearer token).
     */
    fun resume(turnId: String, sessionId: String): Flow<SseEvent>

    /**
     * Explicitly stops a durable turn (`POST /api/chat/turns/:id/stop`; BT-4): the
     * server aborts generation and discards it (nothing persisted). Stopping an
     * unknown or already-terminal turn is a no-op. A failure surfaces as an
     * [OakError]; callers tear down their local stream regardless.
     */
    suspend fun stop(turnId: String, sessionId: String)
}

/**
 * Production [ChatService] over [SseClient] (which borrows `OakApiClient` for the
 * Bearer header + base URL).
 */
class LiveChatService(
    private val sseClient: SseClient,
    private val imageEncoder: ImageEncoder = ImageEncoder(),
) : ChatService {

    override fun send(
        sessionId: String,
        message: String,
        images: List<SourceImage>,
        scopeSeed: Format?,
    ): Flow<SseEvent> {
        // Encode + validate the attached images BEFORE opening the stream. `encode`
        // is synchronous and runs in the caller's context. A client cap/type
        // violation surfaces as a thrown OakError.ImageRejected from the returned
        // flow — never as a partially-attached turn.
        val encodedImages: List<ChatImage> = try {
            imageEncoder.encode(images)
        } catch (e: OakError) {
            return flow { throw e }
        }

        return send(
            ChatRequest(
                sessionId = sessionId,
                message = message,
                images = encodedImages.ifEmpty { null },
                scopeSeed = scopeSeed,
            ),
        )
    }

    override fun send(
        sessionId: String,
        message: String,
        images: List<SourceImage>,
        scopeSeed: Format?,
        recovery: ChatRecovery?,
        mentionedTeamIds: List<String>?,
    ): Flow<SseEvent> {
        val encodedImages: List<ChatImage> = try {
            imageEncoder.encode(images)
        } catch (e: OakError) {
            return flow { throw e }
        }
        return send(
            ChatRequest(
                sessionId = sessionId,
                message = message,
                images = encodedImages.ifEmpty { null },
                scopeSeed = scopeSeed,
                recovery = recovery,
                mentionedTeamIds = mentionedTeamIds?.takeIf { it.isNotEmpty() },
            ),
        )
    }

    override fun send(request: ChatRequest): Flow<SseEvent> = sseClient.stream(request)

    override fun resume(turnId: String, sessionId: String): Flow<SseEvent> =
        sseClient.resume(turnId, sessionId)

    override suspend fun stop(turnId: String, sessionId: String) = sseClient.stop(turnId, sessionId)
}
